src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c

```diff
@@ -413,6 +423,110 @@ static const UCollator* GetCollatorFromSortHandle(SortHandle* pSortHandle, int32
     }
 }
 
+// If succeeded, U_SUCCESS(*pErr) should be true.
+// return -1 if couldn't borrow search handle from the SortHandle cache, otherwise, it return the slot number of the cache.
+static const int32_t GetSearchIteratorFromSortHandleUsingCollator(
+                        SortHandle* pSortHandle,
+                        const UCollator* pColl,
+                        const UChar* lpTarget,
+                        int32_t cwTargetLength,
+                        const UChar* lpSource,
+                        int32_t cwSourceLength,
+                        int32_t options,
+                        UErrorCode* pErr,
+                        UStringSearch** pSearchIterator)
+{
+    options &= CompareOptionsMask;
+    *pSearchIterator = pSortHandle->searchIteratorPerOption[options];
+
+    if (*pSearchIterator == NULL)
+    {
+        *pSearchIterator = usearch_openFromCollator(lpTarget, cwTargetLength, lpSource, cwSourceLength, pColl, NULL, pErr);
+        if (!U_SUCCESS(*pErr))
+        {
+            assert(FALSE && "Couldn't open the search iterator.");
+            return -1;
+        }
+
+        UStringSearch* pNull = NULL;
+        if (!pal_atomic_cas_ptr((void* volatile*)&pSortHandle->searchIteratorPerOption[options], *pSearchIterator, pNull))
+        {
+            usearch_close(*pSearchIterator);
+            *pSearchIterator = pSortHandle->searchIteratorPerOption[options];
+        }
```

Comment:
should we avoid the extra step and just store `USED_STRING_SEARCH` in the cache here and exit immediately on success?

this would avoid the additional calls to `pal_atomic_cas_ptr`, `usearch_setText` and `usearch_setPattern` for the case when the `pSortHandle->searchIteratorPerOption[options]` is `NULL`

```suggestion
        if (!pal_atomic_cas_ptr((void* volatile*)&pSortHandle->searchIteratorPerOption[options], USED_STRING_SEARCH, pNull))
        {
            usearch_close(*pSearchIterator);
            *pSearchIterator = pSortHandle->searchIteratorPerOption[options];
        }
        else
        {
            return options;
        }
```

Reply:
This is possible but we'll have to set the text and the pattern before we return. I can refactor the code to do that.

Reply:
I'll try your suggestion.